src/Microsoft.Data/DataFrameBuffer.cs

```diff
@@ -11,83 +11,57 @@
 namespace Microsoft.Data
 {
     /// <summary>
-    /// A basic store to hold values in a DataFrame column. Supports wrapping with an ArrowBuffer
+    /// A basic immutable store to hold values in a DataFrame column. Supports wrapping with an ArrowBuffer
     /// </summary>
     /// <typeparam name="T"></typeparam>
     public class DataFrameBuffer<T>
         where T : struct
     {
         // TODO: Change this to Memory<T>
-        public Memory<byte> Memory { get; private set; }
 
-        private readonly int _size;
+        private ReadOnlyMemory<byte> _readOnlyMemory;
 
-        private int Capacity => Memory.Length / _size;
+        public virtual ReadOnlyMemory<byte> ReadOnlyMemory => _readOnlyMemory;
 
-        public int MaxCapacity => Int32.MaxValue / _size;
+        protected static int Size = Unsafe.SizeOf<T>();
 
-        public Span<T> Span
+        protected int Capacity => ReadOnlyMemory.Length / Size;
+
+
+        public static int MaxCapacity => Int32.MaxValue / Size;
+
+        public ReadOnlySpan<T> ReadOnlySpan
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => MemoryMarshal.Cast<byte, T>(Memory.Span);
+            get => MemoryMarshal.Cast<byte, T>(ReadOnlyMemory.Span);
         }
 
         public int Length { get; internal set; }
 
         public DataFrameBuffer(int numberOfValues = 8)
         {
-            _size = Unsafe.SizeOf<T>();
-            if ((long)numberOfValues * _size > MaxCapacity)
+            if ((long)numberOfValues * Size > MaxCapacity)
             {
                 throw new ArgumentException($"{numberOfValues} exceeds buffer capacity", nameof(numberOfValues));
             }
-            Memory = new byte[numberOfValues * _size];
+            _readOnlyMemory = new byte[numberOfValues * Size];
         }
 
-        public void Append(T value)
+        public DataFrameBuffer(ReadOnlyMemory<byte> buffer, int length)
```

Comment:
Why do we need a separate `length`. The buffer already has a Length, why can't we use it?

Reply:
Because `length` is the Length of the `DataFrameBuffer`, which can be different from `buffer.Length`. This happens because `DataFrameBuffer` stores everything as `Memory<byte>`. So for ex, the test case `TestArrowStringColumnWithNulls` in BufferTests.cs declares a `DataFrameBuffer<int>` like this:

            Memory<byte> offsetMemory = new byte[] { 0, 0, 0, 0, 3, 0, 0, 0, 3, 0, 0, 0, 7, 0, 0, 0, 7, 0, 0, 0 };
            DataFrameBuffer<int> offsetBuffer = new DataFrameBuffer<int>(offsetMemory, 5);

Reply:
It also occured to me that I could directly wrap an `ArrowBuffer` here instead of deconstructing the `ReadOnlyMemory` from the `ArrowBuffer`. That would remove the issue with specifying a `length` too. Does that seem better?

Reply:
I think it would be best to limit the places that were dependent on Arrow here. If we can keep these buffers general purpose, I think that would be good.

Reply:
Ok, done.